src/query/expression/src/kernels/group_by_hash.rs

```diff
@@ -217,20 +215,38 @@ impl HashMethod for HashMethodSerializer {
     fn build_keys_state(
         &self,
         group_columns: &[(Column, DataType)],
-        rows: usize,
+        num_rows: usize,
     ) -> Result<KeysState> {
-        let approx_size = group_columns.len() * rows * 8;
-        let mut builder = StringColumnBuilder::with_capacity(rows, approx_size);
-
-        for row in 0..rows {
+        let mut data_size = 0;
+        for (column, _) in group_columns.iter() {
+            data_size += column.serialize_size();
+        }
+        let mut data: Vec<u8> = Vec::with_capacity(data_size);
+        let mut offsets: Vec<u64> = Vec::with_capacity(num_rows + 1);
+        offsets.push(0);
+        let mut data_ptr = data.as_mut_ptr();
+        let mut offsets_ptr = unsafe { offsets.as_mut_ptr().add(1) };
```

Comment:
Using unsafe code can lead to hard in maintenance and readability. 
What would be the impact on performance if we don't to employ unsafe code?

Reply:
This code uses pointers to copy data instead of vec push, which does not reduce readability a lot. If we don't use unsafe, we can't advance the pointer. Because the build_keys_state function is a low-level function and we don't modify it very often, it's worth doing this 😊.

Reply:
You can initialize the offsets to be all zeros and write the value without bound check.

```
let offsets = vec![num_rows + 1; 0];

offsets[index] = v;
```

Reply:
`vec![num_rows + 1; 0]` is more expensive than `Vec::with_capacity(num_rows + 1)` because `vec![num_rows + 1; 0]` requires initialization.

Reply:
You can use with_capacity and cast the buffer into &mut [u8] with unsafe code.

```
 offsets.reserve(length + 1);
  let out_slice = unsafe {
      core::slice::from_raw_parts_mut(
          offsets.as_mut_ptr().add(offsets.len()) as *mut u8,
          (length + 1) * std::mem::size_of::<O>(),
      )
  };
```

Reply:
Yes, but the type of `offsets` is `u64`, it is inconvenient to use `&mut [u8]`

Reply:
could change to `*mut u64`

Reply:
Yes, but why not just `unsafe { offsets.as_mut_ptr().add(1) }` to get `*mut u64`.